From eee51352893983a1f509f7477faba6426723afbd Mon Sep 17 00:00:00 2001
From: DogEZ <ddogezd@gmail.com>
Date: Sun, 5 Jul 2026 15:51:16 +0000
Subject: [PATCH] treewide: Add LXC/container support
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Enable kernel features required for Droidspaces/LXC containers:
- Enable SYSVIPC, POSIX_MQUEUE, PID_NS, IPC_NS and other namespace/cgroup configs
- Fix SYSVIPC kABI compatibility by moving sysvsem/sysvshm into ANDROID_KABI area
- Fix overlayfs dentry flags by removing DCACHE_OP_HASH/COMPARE
- Add cgroup noprefix symlink support for container runtimes
- Relax module version check for third-party modules

Credit to CoolAPK @六道纸花
---
 arch/arm64/configs/gki_defconfig | 52 ++++++++++++++++++++++++++++++--
 fs/overlayfs/util.c              |  4 +--
 include/linux/sched.h            | 12 ++++++--
 kernel/cgroup/cgroup.c           |  5 +++
 kernel/module.c                  |  2 +-
 5 files changed, 67 insertions(+), 8 deletions(-)

diff --git a/arch/arm64/configs/gki_defconfig b/arch/arm64/configs/gki_defconfig
index c789ee7f731e..598e218fdb15 100755
--- a/arch/arm64/configs/gki_defconfig
+++ b/arch/arm64/configs/gki_defconfig
@@ -32,7 +32,6 @@ CONFIG_CPUSETS=y
 CONFIG_CGROUP_CPUACCT=y
 CONFIG_CGROUP_BPF=y
 CONFIG_NAMESPACES=y
-# CONFIG_PID_NS is not set
 CONFIG_RT_SOFTINT_OPTIMIZATION=y
 CONFIG_RELAY=y
 # CONFIG_RD_BZIP2 is not set
@@ -41,7 +40,6 @@ CONFIG_RELAY=y
 # CONFIG_RD_LZO is not set
 CONFIG_BOOT_CONFIG=y
 # CONFIG_SYSFS_SYSCALL is not set
-# CONFIG_FHANDLE is not set
 CONFIG_CC_OPTIMIZE_FOR_PERFORMANCE_O3=y
 CONFIG_KALLSYMS_ALL=y
 CONFIG_USERFAULTFD=y
@@ -715,3 +713,53 @@ CONFIG_HIST_TRIGGERS=y
 CONFIG_PID_IN_CONTEXTIDR=y
 CONFIG_FUNCTION_ERROR_INJECTION=y
 # CONFIG_RUNTIME_TESTING_MENU is not set
+CONFIG_POSIX_MQUEUE=y
+CONFIG_PID_NS=y
+CONFIG_SYSVIPC=y
+CONFIG_IPC_NS=y
+CONFIG_CGROUP_DEVICE=y
+CONFIG_BRIDGE_NETFILTER=y
+CONFIG_NF_NAT_IPV4=y
+CONFIG_NETFILTER_XT_MATCH_ADDRTYPE=y
+CONFIG_NF_NAT_NEEDED=y
+CONFIG_USER_NS=y
+CONFIG_CGROUP_PIDS=y
+CONFIG_CFQ_GROUP_IOSCHED=y
+CONFIG_NET_CLS_CGROUP=y
+CONFIG_IP_SCTP=y
+CONFIG_IP_VS=y
+CONFIG_IP_VS_PROTO_TCP=y
+CONFIG_IP_VS_PROTO_UDP=y
+CONFIG_IP_VS_NFCT=y
+CONFIG_VXLAN=y
+CONFIG_MACVLAN=y
+CONFIG_DEVPTS_MULTIPLE_INSTANCES=y
+CONFIG_NETFILTER_XT_MATCH_IPVS=y
+CONFIG_IP_VS_RR=y
+CONFIG_IP6_NF_NAT=y
+CONFIG_IP6_NF_TARGET_MASQUERADE=y
+CONFIG_SECURITY_APPARMOR=y
+CONFIG_NETFILTER_XT_TARGET_CHECKSUM=y
+CONFIG_FHANDLE=y
+CONFIG_UNIX_DIAG=y
+CONFIG_PACKET_DIAG=y
+CONFIG_NETLINK_DIAG=y
+CONFIG_NETFILTER_XT_MATCH_CGROUP=y
+CONFIG_RESOURCE_COUNTERS=y
+CONFIG_IOSCHED_CFQ=y
+CONFIG_CGROUP_PERF=y
+CONFIG_MEMCG_SWAP_ENABLED=y
+CONFIG_XFRM_ALGO=y
+CONFIG_AUFS_FS=y
+CONFIG_BTRFS_FS=y
+CONFIG_BTRFS_FS_POSIX_ACL=y
+CONFIG_DM_THIN_PROVISIONING=y
+CONFIG_NF_CONNMARK_IPV4=y
+CONFIG_NF_NAT_MASQUERADE_IPV4=y
+CONFIG_CGROUP_NET_CLASSID=y
+CONFIG_PERSISTENT_KEYRINGS=y
+CONFIG_ENCRYPTED_KEYS=y
+CONFIG_KEY_DH_OPERATIONS=y
+#CONFIG_NET_L3_MASTER_DEV is not set
+CONFIG_BRIDGE_VLAN_FILTERING=y
+CONFIG_IPVLAN=y
\ No newline at end of file
diff --git a/fs/overlayfs/util.c b/fs/overlayfs/util.c
index f96713782c33..543241e9a9a0 100644
--- a/fs/overlayfs/util.c
+++ b/fs/overlayfs/util.c
@@ -148,9 +148,7 @@ bool ovl_dentry_weird(struct dentry *dentry)
 		return true;
 
 	return dentry->d_flags & (DCACHE_NEED_AUTOMOUNT |
-				  DCACHE_MANAGE_TRANSIT |
-				  DCACHE_OP_HASH |
-				  DCACHE_OP_COMPARE);
+				  DCACHE_MANAGE_TRANSIT);
 }
 
 enum ovl_path_type ovl_path_type(struct dentry *dentry)
diff --git a/include/linux/sched.h b/include/linux/sched.h
index b275cd285182..dd3101f7963e 100644
--- a/include/linux/sched.h
+++ b/include/linux/sched.h
@@ -1074,8 +1074,8 @@ struct task_struct {
 	struct nameidata		*nameidata;
 
 #ifdef CONFIG_SYSVIPC
-	struct sysv_sem			sysvsem;
-	struct sysv_shm			sysvshm;
+//	struct sysv_sem			sysvsem;
+//	struct sysv_shm			sysvshm;
 #endif
 #ifdef CONFIG_DETECT_HUNG_TASK
 	unsigned long			last_switch_count;
@@ -1508,9 +1508,17 @@ struct task_struct {
 	struct callback_head		l1d_flush_kill;
 #endif
 
+#if defined(CONFIG_SYSVIPC)
+	// struct sysv_sem			sysvsem;
+	ANDROID_KABI_USE(1, struct sysv_sem sysvsem);
+	// struct sysv_shm			sysvshm;
+	_ANDROID_KABI_REPLACE(ANDROID_KABI_RESERVE(2); ANDROID_KABI_RESERVE(3),
+						  struct sysv_shm sysvshm);
+#else
 	ANDROID_KABI_RESERVE(1);
 	ANDROID_KABI_RESERVE(2);
 	ANDROID_KABI_RESERVE(3);
+#endif
 	ANDROID_KABI_RESERVE(4);
 	ANDROID_KABI_RESERVE(5);
 	ANDROID_KABI_RESERVE(6);
diff --git a/kernel/cgroup/cgroup.c b/kernel/cgroup/cgroup.c
index b3d5407405b4..f45d35b7323d 100644
--- a/kernel/cgroup/cgroup.c
+++ b/kernel/cgroup/cgroup.c
@@ -4132,6 +4132,11 @@ static int cgroup_add_file(struct cgroup_subsys_state *css, struct cgroup *cgrp,
 		spin_unlock_irq(&cgroup_file_kn_lock);
 	}
 
+	if (cft->ss && (cgrp->root->flags & CGRP_ROOT_NOPREFIX) && !(cft->flags & CFTYPE_NO_PREFIX)) {
+		snprintf(name, CGROUP_FILE_NAME_MAX, "%s.%s", cft->ss->name, cft->name);
+		kernfs_create_link(cgrp->kn, name, kn);
+	}
+
 	return 0;
 }
 
diff --git a/kernel/module.c b/kernel/module.c
index bcadcef319dc..3794f311ba02 100644
--- a/kernel/module.c
+++ b/kernel/module.c
@@ -1289,7 +1289,7 @@ static int check_version(const struct load_info *info,
 bad_version:
 	pr_warn("%s: disagrees about version of symbol %s\n",
 	       info->name, symname);
-	return 0;
+	return 1;
 }
 
 static inline int check_modstruct_version(const struct load_info *info,
-- 
2.53.0

